test/integration/ctl/ctl_create_cr_test.go

```diff
@@ -0,0 +1,240 @@
+			expName:        cr1Name,
+			expKeyFilename: cr1Name + ".key",
+		},
+		"v1alpha3 Certificate given": {
+			inputFile:      testdataPath + "create_cr_v1alpha3_cert_with_ns1.yaml",
+			inputArgs:      []string{cr2Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expValidateErr: false,
+			expRunErr:      false,
+			expNamespace:   ns1,
+			expName:        cr2Name,
+			expKeyFilename: cr2Name + ".key",
+		},
+		"conflicting namespaces defined in flag and file": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr3Name},
+			inputNamespace: ns2,
+			keyFilename:    "",
+			expValidateErr: false,
+			expRunErr:      true,
+			expNamespace:   "",
+			expName:        "",
+			expKeyFilename: "",
+		},
+		"file passed in defines resource other than certificate": {
+			inputFile:      testdataPath + "create_cr_issuer.yaml",
+			inputArgs:      []string{cr4Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expValidateErr: false,
+			expRunErr:      true,
+			expNamespace:   "",
+			expName:        "",
+			expKeyFilename: "",
+		},
+		"path to file to store private key provided": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr5Name},
+			inputNamespace: ns1,
+			keyFilename:    "test.key",
+			expValidateErr: false,
+			expRunErr:      false,
+			expNamespace:   ns1,
+			expName:        cr5Name,
+			expKeyFilename: "test.key",
+		},
+		"CR name not passed as arg": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expValidateErr: true,
+			expRunErr:      false,
+			expNamespace:   ns1,
+			expKeyFilename: "",
+		},
+	}
+
+	for name, test := range tests {
+		// Run ctl create cr command with input options
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			// Options to run create CR command
+			opts := &certificaterequest.Options{
+				CMClient:         cmCl,
+				RESTConfig:       config,
+				IOStreams:        streams,
+				CmdNamespace:     test.inputNamespace,
+				EnforceNamespace: test.inputNamespace != "",
+				KeyFilename:      test.keyFilename,
+			}
+
+			opts.InputFilename = test.inputFile
+
+			// Validating args and flags
+			err := opts.Validate(test.inputArgs)
+			if err != nil {
+				if !test.expValidateErr {
+					t.Errorf("got unexpected error when validating args and flags: %v", err)
+				}
+				t.Logf("got an error, which was expected, details: %v", err)
+				return
+			} else {
+				// got no error
+				if test.expValidateErr {
+					t.Errorf("expected but got no error validating args and flags")
+				}
+			}
+
+			// Create CR
+			err = opts.Run(test.inputArgs)
+			if err != nil {
+				if !test.expRunErr {
+					t.Errorf("got unexpected error when trying to create CR: %v", err)
+				}
+				t.Logf("got an error, which was expected, details: %v", err)
+				return
```

Comment:
In cases where creating the *CertificateRequest* fails, but not writing the private key, this return will cause us to not perform any later checks/test code for to ensure that the private key at least has been written correctly. Is this intended/expected? If so, can you add a comment here just to note it, and note that we may want to change this in future? 😄